pandas/core/reshape.py

```diff
@@ -202,10 +202,13 @@ def get_new_columns(self):
                 return self.removed_level
 
             lev = self.removed_level
-            vals = np.insert(lev.astype('object'), 0,
+            vals = np.insert(lev.values.astype('object'), 0,
                              _get_na_value(lev.dtype.type))
 
-            return lev._shallow_copy(vals)
+            try:
+                return lev._shallow_copy(vals.astype(lev.dtype, copy=False))
+            except ValueError:
```

Comment:
`_shallow_copy` does not do any type inference at all; `Index` does inspect types quite a lot, in particular [this line](https://github.com/pydata/pandas/blob/02cbd9e0f3e683eb95e74dfc0e8ba1bd2ad29872/pandas/core/index.py#L195) is very inefficient. if i do `Index(vals.astype(lev.dtype, copy=False), ...` or `Index(vals, dtype=lev.dtype)` it would still need a `try ... catch` block because of the possible `ValueError`